### PR TITLE
Add mypy to pre-commit

### PR DESCRIPTION
**Pull request by Dhwaniartefact**
This PR introduces `mypy` in the `pre-commit` configurations and adds typing annotations in `test_fixity` tests.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,3 +21,11 @@ repos:
       (?x)^(
         (README)\.md
       )
+- repo: https://github.com/pre-commit/mirrors-mypy
+  rev: v1.10.1
```
Comment (replaceafill) [inline .pre-commit-config.yaml]:
```suggestion
  rev: v1.11.2
```

File: fixity/models.py [hunk 2]
```diff
@@ -16,7 +18,7 @@
 engine = create_engine(f"sqlite:///{db_path}", echo=False)
 
 Session = sessionmaker(bind=engine)
-Base = declarative_base()
+Base: Type[DeclarativeBase] = declarative_base()
```
Comment (replaceafill) [inline fixity/models.py]: I think there's a more convenient ways to type this. See the `Base` class used in https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#step-one-orm-declarative-base-is-superseded-by-orm-declarativebase

File: pyproject.toml [hunk 3]
```diff
@@ -127,3 +127,20 @@ legacy_tox_ini = """
...
+    "tests.*",
+]
+ignore_errors = true
+
+[[tool.mypy.overrides]]
+module = [
+    "tests.test_fixity",
+]
+ignore_errors = false
+disallow_untyped_calls = false
```
Comment (replaceafill) [inline pyproject.toml]: Why is `disallow_untyped_calls = false` necessary here?

Also this file needs a new line at the end. See the marker GitHub displays as an indication for that.

File: tests/test_fixity.py [hunk 4]
```diff
@@ -1,8 +1,10 @@
 import io
 import json
+import typing
 import uuid
 from datetime import datetime
 from datetime import timezone
+from typing import List
```
Comment (replaceafill) [inline tests/test_fixity.py]: Could we stick with a consistent way to import from the `typing` module?

This is importing the whole module at the top and then a class from it at the bottom.

File: tests/test_fixity.py [hunk 5]
```diff
@@ -54,13 +56,15 @@ def mock_check_fixity():
...
-def _assert_stream_content_matches(stream, expected):
+def _assert_stream_content_matches(stream: typing.TextIO, expected: List[str]) -> None:
     stream.seek(0)
     assert [line.strip() for line in stream.readlines()] == expected
 
 
 @mock.patch("requests.get")
-def test_scan(_get, environment, mock_check_fixity):
+def test_scan(
+    _get: mock.Mock, environment: mock.Mock, mock_check_fixity: mock.Mock
```
Comment (replaceafill) [inline tests/test_fixity.py]: Notice the type hints you used for the `environment` and `mock_check_fixity` above. You should use the same type hints here, i.e. `environment: None, mock_check_fixity: List[mok.Mock]`.

File: tests/test_fixity.py [hunk 6]
```diff
@@ -86,8 +90,11 @@ def test_scan(_get, environment, mock_check_fixity):
 @mock.patch("time.time")
 @mock.patch("requests.get")
 def test_scan_if_timestamps_argument_is_passed(
-    _get, time, environment, mock_check_fixity
-):
+    _get: mock.Mock,
+    time: mock.Mock,
+    environment: pytest.MonkeyPatch,
+    mock_check_fixity: mock.Mock,
```
Comment (replaceafill) [inline tests/test_fixity.py]: Notice the type hints you used for the `environment` and `mock_check_fixity` above. You should use the same type hints here, i.e. `environment: None, mock_check_fixity: List[mok.Mock]`.

File: tests/test_fixity.py [hunk 7]
```diff
@@ -126,8 +133,14 @@ def test_scan_if_timestamps_argument_is_passed(
...
 )
 def test_scan_if_report_url_exists(
-    _post, _get, utcnow, uuid4, mock_check_fixity, environment, monkeypatch
-):
+    _post: mock.Mock,
+    _get: mock.Mock,
+    utcnow: mock.Mock,
+    uuid4: mock.Mock,
+    mock_check_fixity: mock.Mock,
+    environment: pytest.MonkeyPatch,
```
Comment (replaceafill) [inline tests/test_fixity.py]: Notice the type hints you used for the `environment` and `mock_check_fixity` above. You should use the same type hints here, i.e. `environment: None, mock_check_fixity: List[mok.Mock]`.

Also it'd be nice if you could keep the order of the fixtures consistent across tests.

File: tests/test_fixity.py [hunk 8]
```diff
@@ -197,8 +210,12 @@ def test_scan_if_report_url_exists(
     ],
 )
 def test_scan_handles_exceptions_if_report_url_exists(
-    _post, _get, environment, monkeypatch, mock_check_fixity
-):
+    _post: mock.Mock,
+    _get: mock.Mock,
+    environment: pytest.MonkeyPatch,
+    monkeypatch: pytest.MonkeyPatch,
+    mock_check_fixity: mock.Mock,
```
Comment (replaceafill) [inline tests/test_fixity.py]: Notice the type hints you used for the `environment` and `mock_check_fixity` above. You should use the same type hints here, i.e. `environment: None, mock_check_fixity: List[mok.Mock]`.

Also it'd be nice if you could keep the order of the fixtures consistent across tests.

File: tests/test_fixity.py [hunk 9]
```diff
@@ -237,7 +254,9 @@ def test_scan_handles_exceptions_if_report_url_exists(
         ),
     ],
 )
-def test_scan_handles_exceptions(_get, environment):
+def test_scan_handles_exceptions(
+    _get: mock.Mock, environment: pytest.MonkeyPatch
```
Comment (replaceafill) [inline tests/test_fixity.py]:
```suggestion
    _get: mock.Mock, environment: None
```

File: tests/test_fixity.py [hunk 10]
```diff
@@ -272,7 +291,9 @@ def test_scan_handles_exceptions(_get, environment):
         ),
     ],
 )
-def test_scan_handles_exceptions_if_no_scan_attempted(_get, environment):
+def test_scan_handles_exceptions_if_no_scan_attempted(
+    _get: mock.Mock, environment: pytest.MonkeyPatch
```
Comment (replaceafill) [inline tests/test_fixity.py]:
```suggestion
    _get: mock.Mock, environment: None
```

File: tests/test_fixity.py [hunk 11]
```diff
@@ -291,7 +312,7 @@ def test_scan_handles_exceptions_if_no_scan_attempted(_get, environment):
     ],
     ids=["Success", "Fail", "Did not run"],
 )
-def test_scan_message(status, error_message):
+def test_scan_message(status: mock.Mock, error_message: mock.Mock) -> None:
```
Comment (replaceafill) [inline tests/test_fixity.py]: These parameters are not `Mock` objects. See https://docs.pytest.org/en/latest/example/parametrize.html and the reference of the `parametrize` mark https://docs.pytest.org/en/latest/reference/reference.html#pytest.Metafunc.parametrize

The `argvalues` in this test a couple lines above are tuples of `(bool, str)`. That's what the test will receive.

File: tests/test_fixity.py [hunk 12]
```diff
@@ -306,7 +327,9 @@ def test_scan_message(status, error_message):
 @mock.patch(
     "requests.get",
 )
-def test_scanall(_get, environment, mock_check_fixity):
+def test_scanall(
+    _get: mock.Mock, environment: pytest.MonkeyPatch, mock_check_fixity: mock.Mock
```
Comment (replaceafill) [inline tests/test_fixity.py]: Same comment as before.

File: tests/test_fixity.py [hunk 13]
```diff
@@ -351,7 +374,9 @@ def test_scanall(_get, environment, mock_check_fixity):
 
 
 @mock.patch("requests.get")
-def test_scanall_handles_exceptions(_get, environment):
+def test_scanall_handles_exceptions(
+    _get: mock.Mock, environment: pytest.MonkeyPatch
```
Comment (replaceafill) [inline tests/test_fixity.py]:
```suggestion
    _get: mock.Mock, environment: None
```

File: tests/test_fixity.py [hunk 14]
```diff
@@ -412,7 +437,9 @@ def test_scanall_handles_exceptions(_get, environment):
 
 
 @mock.patch("requests.get")
-def test_main_handles_exceptions_if_scanall_fails(_get, environment):
+def test_main_handles_exceptions_if_scanall_fails(
+    _get: mock.Mock, environment: pytest.MonkeyPatch
```
Comment (replaceafill) [inline tests/test_fixity.py]:
```suggestion
    _get: mock.Mock, environment: None
```

File: tests/test_fixity.py [hunk 15]
```diff
@@ -473,7 +500,9 @@ def test_main_handles_exceptions_if_scanall_fails(_get, environment):
 
 
 @mock.patch("requests.get")
-def test_scanall_if_sort_argument_is_passed(_get, environment, mock_check_fixity):
+def test_scanall_if_sort_argument_is_passed(
+    _get: mock.Mock, environment: pytest.MonkeyPatch, mock_check_fixity: mock.Mock
```
Comment (replaceafill) [inline tests/test_fixity.py]: Same comment as before.

**Review by replaceafill — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 3 above)
Comment (Dhwaniartefact) [reply pyproject.toml]: I have added this due to https://mypy.readthedocs.io/en/stable/command_line.html#untyped-definitions-and-calls

Alternatively, I can add `# type: ignore` in the beginning of the file as I dont want to apply annotations in `fixity.py` yet.

Comment (replaceafill) [reply pyproject.toml]: Aren't you ignoring the `fixity` module with:

https://github.com/artefactual/fixity/blob/61b27917890c8175e5a6e31bf6a5bdf7d9804109/pyproject.toml#L134-L139

Sorry, I still don't understand why the setting is necessary when:

1. You're setting it to its [default value](https://mypy.readthedocs.io/en/stable/config_file.html#confval-disallow_untyped_calls).
2. You're looking for strict validation in the test which the default value doesn't accomplish. From the `--strict` flag on `mypy --help`:
    ```console
      --strict              Strict mode; enables the following flags: --warn-
                            unused-configs, --disallow-any-generics,
                            --disallow-subclassing-any, --disallow-untyped-
                            calls, --disallow-untyped-defs, --disallow-
                            incomplete-defs, --check-untyped-defs, --disallow-
                            untyped-decorators, --warn-redundant-casts,
                            --warn-unused-ignores, --warn-return-any, --no-
                            implicit-reexport, --strict-equality, --extra-
                            checks
    ```

File: fixity/fixity.py [hunk 16]
```diff
@@ -305,7 +309,9 @@ def get_handler(stream, timestamps, log_level=None):
     return handler
 
 
-def main(argv=None, logger=None, stream=None):
+def main(
+    argv: List[str] = None, logger: logging.Logger = None, stream: TextIO = None
```
Comment (replaceafill) [inline fixity/fixity.py]: I think all these parameters that default to `None` should be marked as [Optional](https://docs.python.org/3.8/library/typing.html#typing.Optional).

File: fixity/fixity.py [hunk 17]
```diff
@@ -305,7 +309,9 @@ def get_handler(stream, timestamps, log_level=None):
...
+) -> Union[int, bool]:
```
Comment (replaceafill) [inline fixity/fixity.py]: Have you seen the `try... except` blocks inside this function? Don't they also return exceptions?

File: fixity/fixity.py [hunk 18]
```diff
@@ -91,7 +95,7 @@ def fetch_environment_variables(namespace):
         namespace.report_user = namespace.report_pass = None
 
 
-def scan_message(aip_uuid, status, message):
+def scan_message(aip_uuid: uuid.UUID, status: bool, message: str) -> str:
```
Comment (replaceafill) [inline fixity/fixity.py]: Why is `aip_uuid` typed as a `UUID`? The only call to `scan_message` I see is:

https://github.com/artefactual/fixity/blob/55e50db590ed937bd06553f56a1c0e13bed0d727/fixity/fixity.py#L171

And the docstring for `scan` documents it as string:

https://github.com/artefactual/fixity/blob/55e50db590ed937bd06553f56a1c0e13bed0d727/fixity/fixity.py#L128

Am I missing something?

File: fixity/models.py [hunk 19]
```diff
@@ -7,16 +7,19 @@
...
 
 db_path = os.path.join(os.path.dirname(__file__), "fixity.db")
 engine = create_engine(f"sqlite:///{db_path}", echo=False)
 
 Session = sessionmaker(bind=engine)
-Base = declarative_base()
+
+
+class Base(DeclarativeBase):
+    pass
```
Comment (replaceafill) [inline fixity/models.py:22]: :tada:

**Review by replaceafill — CHANGES_REQUESTED** [review]

File: tests/test_fixity.py (hunk 7 above)
Comment (replaceafill) [reply tests/test_fixity.py]: > Also it'd be nice if you could keep the order of the fixtures consistent across tests.

Could you please address this?

File: tests/test_fixity.py (hunk 8 above)
Comment (replaceafill) [reply tests/test_fixity.py]: > Also it'd be nice if you could keep the order of the fixtures consistent across tests.

Could you please address this?

File: fixity/fixity.py (hunk 18 above)
Comment (Dhwaniartefact) [reply fixity/fixity.py]: Nope, I need to change it.

File: fixity/fixity.py (hunk 17 above)
Comment (Dhwaniartefact) [reply fixity/fixity.py]: oh yes, I skipped this.

**Review by replaceafill — APPROVED** [review]

**Pull request merged**